vtkext/private/module/vtkF3DImguiConsole.cxx

```diff
@@ -14,13 +17,153 @@ struct vtkF3DImguiConsole::Internals
+    {
+      case ImGuiInputTextFlags_CallbackCompletion:
+      {
+        assert(this->GetCommandsMatchCallback);
+        std::string pattern{ data->Buf };
+        std::vector<std::string> candidates =
+          this->GetCommandsMatchCallback(pattern); // List of supported commands
+
+        if (candidates.size() == 1)
+        {
+          // Single match. Delete the beginning of the word and replace it entirely so we've got
+          // nice casing.
+          data->DeleteChars(0, static_cast<int>(pattern.size()));
+          data->InsertChars(data->CursorPos, candidates[0].c_str());
+          data->InsertChars(data->CursorPos, " ");
+        }
+        else if (candidates.size() > 1)
+        {
+          // Multiple matches. Complete as much as we can.
+          // So inputting "C"+Tab will complete to "CL" then display "CLEAR" and "CLASSIFY" as
+          // matches.
+          size_t matchLen = pattern.size();
+          bool allCandidatesMatches = true;
+          // Find the common prefix to all candidates
+          while (allCandidatesMatches)
+          {
+            const std::string& first = candidates[0];
+            if (first.size() <= matchLen)
+            {
+              // The first candidate is shorter than the current match length
+              allCandidatesMatches = false;
+            }
+            else
+            {
+              // Check if all candidates match the current character
+              const char target = first[matchLen];
+              allCandidatesMatches = std::all_of(candidates.begin(), candidates.end(),
+                [matchLen, target](const std::string& s)
+                { return s.size() > matchLen && s[matchLen] == target; });
+            }
+            if (allCandidatesMatches)
+            {
+              matchLen++;
+            }
+          }
+
+          if (matchLen > 0)
+          {
+            // Fill the best we can by now - use the longest common prefix from available candidates
+            // (possibly just pattern itself in the worst case)
+            data->DeleteChars(0, static_cast<int>(pattern.size()));
+            data->InsertChars(
+              data->CursorPos, candidates[0].c_str(), candidates[0].c_str() + matchLen);
+          }
+
+          this->Completions.first = this->Logs.size();
+          this->Completions.second = this->Logs.size() + candidates.size() + 1;
+          // Add all candidates to the logs
+          this->Logs.emplace_back(
+            std::make_pair(Internals::LogType::Completion, "Possible matches:"));
+          std::transform(candidates.begin(), candidates.end(), std::back_inserter(this->Logs),
+            [](const std::string& candidate)
+            { return std::make_pair(Internals::LogType::Completion, candidate); });
+        }
+        break;
+      }
+      case ImGuiInputTextFlags_CallbackHistory:
+      {
+        const int prevHistoryPos = this->CommandHistoryIndexInv;
+        if (data->EventKey == ImGuiKey_UpArrow)
```

Comment:
add a comment explainting that CommandHistoryIndexInv move between [-1, this->CommandHistory.size()] with up/down key ?